### ISSUE TITLE
frontmatter is targeted

### ISSUE THREAD

**Issue by chenxsan** [mdx-js/eslint-mdx#336]
### Subject of the issue

Say I have a MDX file with this frontmatter:

<pre>
---
title: Code Splitting
related:
  - title: &lt;link rel="prefetch/preload" > in webpack
    url: https://medium.com/webpack/link-rel-prefetch-preload-in-webpack-51a52358f84c
---
</pre>
Note that the `<link >` was not closed.

Now eslint would report `Parsing error: Expected corresponding JSX closing tag for <link>`. Which seems wrong to me. A related comment is here https://github.com/webpack/webpack.js.org/pull/5312#issuecomment-902705944.

### Your environment

*   **OS**: 
*   **Packages**: 
*   **Env**: 

### Steps to reproduce

1. clone https://github.com/webpack/webpack.js.org
2. open `src/content/guides/code-splitting.mdx` and replace the `/>` with `>` in the related part of frontmatter.

FYI, I've tried the latest version https://github.com/mdx-js/eslint-mdx/issues/334#issuecomment-898999220 with no luck.

### Expected behaviour

No error.

### Actual behaviour

An error was shown because frontmatter is targetted.

**Comment by JounQin** [comment]
`mdx` is `jsx` like.

**Comment by chenxsan** [comment]
I'm not familiar with all those remark/eslint stuffs.

But as far as I know, the frontmatter would be transformed into something like:

```js
export const frontmatter = {
  title: '<link >'
}
```
Does it really make sense to lint that part?

**Comment by JounQin** [comment]
That would depend on `remark-frontmatter`'s behavior.

@wooorm Will the following work as same as above?

```mdx
---
title: Code Splitting
related:
  - title: <link rel="prefetch/preload" /> in webpack
    url: <https://medium.com/webpack/link-rel-prefetch-preload-in-webpack-51a52358f84c>
---
```

**Comment by atesgoral** [comment]
Shouldn't the frontmatter be preprocessed and put aside as metadata before the JSX parser is unleashed on the rest of the document?

**Comment by JounQin** [comment]
> Shouldn't the frontmatter be preprocessed and put aside as metadata before the JSX parser is unleashed on the rest of the document?

I'm not familiar with frontmatter, let's wait @wooorm to clarify first. Personally, I don't want to add specific logic for a specific plugin.

If you have any good thought, welcome to PR to preview first.

**Comment by wooorm** [comment]
Frontmatter, often in YAML, is frequently used in combination with markdown/MDX: https://github.com/wooorm/xdm#frontmatter. While ESM import/exports are powerful, frontmatter is static: it can be access without compiling and evaluating the file.
GitHub, prettier, and many blog platforms support it. [mdx-bundler does, too](https://github.com/kentcdodds/mdx-bundler#frontmatter-and-const).

---

> But as far as I know, the frontmatter would be transformed into something like:
>
> ```js
> export const frontmatter = {
>   title: '<link >'
> }
> ```
> — @chenxsan

Not necessarily, frontmatter has always worked with a plugin, but exposing it as an `export` only started working since xdm, and only when combined with [`remcohaszing/remark-mdx-frontmatter`](https://github.com/remcohaszing/remark-mdx-frontmatter). But it’s really smart and powerful so I can understand it becoming quite popular.

> Will the following work as same as above?
> — @JounQin

If `remark-frontmatter` is turned on (with YAML, which is the default), then yes, that parses YAML frontmatter, not as JSX.

> It's parsed as jsx in paragraph even used with remark-frontmatter, that's the root cause.
> — @JounQin

It sounds like `remark-frontmatter` isn’t used correctly in your test case? It should work: https://astexplorer.net/#/gist/85406af039218fbe83c53f831da662be/244e40f3f5208d9c11922e8b3885c5095afbe0a7 (this example is just remark, but I’m pretty sure MDX parsing doesn’t conflict)

> Shouldn't the frontmatter be preprocessed and put aside as metadata before the JSX parser is unleashed on the rest of the document?
> — @atesgoral

Yes, I agree. frontmatter in YAML is really strict. It is highly unlikely that someone *does not* want frontmatter but still starts a document with those three dashes.
Frontmatter is pretty simple to parse (compared to say, JSX), so there shouldn’t be much added code weight either.
Alternatively, it could make it an option.

**Comment by JounQin** [comment]
@wooorm 

> It sounds like remark-frontmatter isn’t used correctly in your test case?

It is using `remark@13` at AST explorer, which is not supported by `eslint-mdx` currently.

My test case is `remark-frontmatter@2`, is this related?

**Comment by wooorm** [comment]
I believe `remark-frontmatter@2` works with `remark@12`. `remark-frontmatter@3` and `remark-frontmatter@4` work with `remark@13` and `remark@14`? 🤔

**Comment by JounQin** [comment]
@wooorm I mean will the parsed AST be same between `remark-frontmatter@2` vs `remark-frontmatter@3/4`?

**Comment by wooorm** [comment]
Yes, the AST is the same with `remark-frontmatter` 1 through 4, and most likely all future versions as well. The MDX AST did you change. But I do not see that change interfering with frontmatter.

**Comment by JounQin** [comment]
Does the plugins order matter? In `eslint-mdx` use `remark-mdx` first then other user customed plugins.

**Comment by JounQin** [comment]
@chenxsan Please try `yarn add -D https://pkg.csb.dev/mdx-js/eslint-mdx/commit/478660c5/eslint-plugin-mdx`

**Comment by chenxsan** [comment]
![image](https://user-images.githubusercontent.com/1091472/130316417-151673b5-0b28-4c9d-8418-4ab22ee756fd.png)

I'm still getting the error.

**Comment by JounQin** [comment]
Did you add `remark-frontmatter@2` in your remark config?

**Comment by chenxsan** [comment]
> @chenxsan Did you add `remark-frontmatter@2` in your remark config?

Oops, I just noticed that I have to set`remarkConfig` in `package.json`. Yes, it works with `remark-frontmatter@2`, but not with v3.

**Comment by JounQin** [comment]
@chenxsan 

See https://github.com/mdx-js/eslint-mdx/issues/336#issuecomment-903066561, `remark-frontmatter@2` is required.
